### DISCUSSION TITLE
MicroPython - Converting to / from timestamp.

### DISCUSSION THREAD

**Discussion by britesc** [micropython/micropython#14243] [Libraries & Drivers]
Hi,
Can anyone please advise me as to how to make a timestamp and conversely convert from a timestamp using MicroPython?
Thanks and kind regards,
jB

**Comment by Rumidom** [comment]
try: time.localtime(timestamp) https://docs.micropython.org/en/latest/library/time.html
it returns a tuple (year, month, mday, hour, minute, second, weekday, yearday)
however some on boards its not Unix time, on esp32 s3 I get:
```python
>>> time.localtime(0)
(2000, 1, 1, 0, 0, 0, 5, 1)
```
to convert back use:
```python
>>> time.mktime((2000, 1, 1, 0, 0, 0, 5, 1))
0
```

**Comment by ned-pcs** [comment]
You don't say what you mean by a "timestamp". But if you mean an ASCII string representing a date/time, you can use the `time.strftime()` from `micropython-lib` to make a printable timestamp.

Parsing one of those is more involved, especially if you have to deal with arbitrary formats.

Here's a function I wrote to make ISO8601 timestamps:
```
import time
from micropython import const

# Allow for unix micropython port which doesn't have RTC
rtc = None
try:
    rtc = machine.RTC()
except AttributeError:
    pass

_ISO_FORMAT_STRING = const("{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}.{:03d}Z")

def timestamp_iso8601(time_secs=None):
    if time_secs is None:
        if rtc is None:
            return timestamp_iso8601(time.time())

        # (year, month, day, weekday, hours, minutes, seconds, subseconds)
        time_tuple = rtc.datetime()
        return _ISO_FORMAT_STRING.format(
            time_tuple[0],
            time_tuple[1],
            time_tuple[2],
            time_tuple[4],
            time_tuple[5],
            time_tuple[6],
            int(time_tuple[7] / 1000),
        )
    else:
        # (year, month, mday, hour, minute, second, weekday, yearday)
        time_tuple = time.localtime(time_secs)
        return _ISO_FORMAT_STRING.format(
            time_tuple[0],
            time_tuple[1],
            time_tuple[2],
            time_tuple[3],
            time_tuple[4],
            time_tuple[5],
            0,
        )
```

**Reply by britesc** [reply]
@ned-pcs 
Hi, 
Thank you for your very exhaustive reply. It will be really useful.
The timestamp I was referring to is the Unix timestamp such as 1712339452 which one usually receives from a NTP service. Which is what I am hoping to achieve.
Once again, many thanks and kind regards,
jB

**Reply by bikeNomad** [reply]
You will have to correct for differing epochs. Apparently the NTP timestamp is relative to the start of the year 1900, the Unix timestamp relative to 1970, and the Micropython timestamp is relative to 2000.
See https://github.com/micropython/micropython-lib/blob/master/micropython/net/ntptime/ntptime.py for details.

**Reply by mendenm** [reply]
@britesc @bikeNomad I think it is even slight more complicated than the y2k base.  The mpy epoch is SDK-dependent for different platforms, so always checking the epoch by doing `time.localtime(0)` is the right (or at least safe) thing.
